Composite/AspNet/CmsPageHttpHandler.cs

```diff
@@ -0,0 +1,147 @@
+                var functionContext = PageRenderer.GetPageRenderFunctionContextContainer();
+
+                XDocument document;
+                DonutCacheEntry cacheEntry = null;
+                if (cachingEnabled)
+                {
+                    using (Profiler.Measure("Cache lookup"))
+                    {
+                        cacheEntry = OutputCacheHelper.GetFromCache(context, cacheKey);
+                    }
+                }
+
+                bool allFunctionsExecuted = false;
+                bool preventResponseCaching = false;
+
+                if (cacheEntry != null)
+                {
+                    document = cacheEntry.Document;
+                    foreach (var header in cacheEntry.OutputHeaders)
+                    {
+                        context.Response.Headers[header.Name] = header.Value;
+                    }
+
+                    // Making sure this response will not go to the output cache
+                    preventResponseCaching = true;
+                }
+                else
+                {
+                    if (renderingContext.RunResponseHandlers())
+                    {
+                        return;
+                    }
+
+                    var renderer = PageTemplateFacade.BuildPageRenderer(renderingContext.Page.TemplateId);
+
+                    var slimRenderer = (ISlimPageRenderer) renderer;
+
+                    using (Profiler.Measure($"{nameof(ISlimPageRenderer)}.Render"))
+                    {
+                        document = slimRenderer.Render(renderingContext.PageContentToRender, functionContext);
+                    }
+                    
+                    allFunctionsExecuted = PageRenderer.ExecuteCacheableFunctions(document.Root, functionContext);
+
+                    if (cachingEnabled && !allFunctionsExecuted && OutputCacheHelper.ResponseCachebale(context))
+                    {
+                        preventResponseCaching = true;
+
+                        if (!functionContext.ExceptionsSuppressed)
+                        {
+                            using (Profiler.Measure("Adding to cache"))
+                            {
+                                OutputCacheHelper.AddToCache(context, cacheKey, new DonutCacheEntry(context, document));
+                            }
+                        }
+                    }
+                }
+
+                if (!allFunctionsExecuted)
+                {
+                    using (Profiler.Measure("Executing embedded functions"))
+                    {
+                        PageRenderer.ExecuteEmbeddedFunctions(document.Root, functionContext);
+                    }
+                }
+
+                using (Profiler.Measure("Resolving page fields"))
+                {
+                    PageRenderer.ResolvePageFields(document, renderingContext.Page);
+                }
+
+                string xhtml;
+                if (document.Root.Name == RenderingElementNames.Html)

```

Comment:
I think it is a false alarm from ReSharper - the only way the null reference would appear, is if PageRenderer.ResolvePageFields remove it, which should be unlikely. It would be more confusing for a developer to see that null check there.